src/crimson/os/seastore/cached_extent.h

```diff
@@ -1385,18 +1387,33 @@ class LBAMapping;
  */
 class LogicalCachedExtent : public CachedExtent {
 public:
-  template <typename... T>
-  LogicalCachedExtent(T&&... t) : CachedExtent(std::forward<T>(t)...) {}
+  using CachedExtent::CachedExtent;
+
+  LogicalCachedExtent(const LogicalCachedExtent& other)
+    : CachedExtent(other),
+     seen_by_users(other.seen_by_users) {}
 
   void on_rewrite(Transaction &t, CachedExtent &extent, extent_len_t off) final {
     assert(get_type() == extent.get_type());
     auto &lextent = (LogicalCachedExtent&)extent;
     set_laddr((lextent.get_laddr() + off).checked_to_laddr());
+    seen_by_users = lextent.seen_by_users;
     do_on_rewrite(t, lextent);
   }
 
   virtual void do_on_rewrite(Transaction &t, LogicalCachedExtent &extent) {}
 
+  bool is_seen_by_users() const {
+    return seen_by_users;
+  }
+
+  // handled under TransactionManager,
+  // user should not set this by themselves.
+  void set_seen_by_users() {
```

Comment:
Maybe we can make TransactionManager a friend of LogicalCachedExtent, so that we can avoid adding this public method which may be wrongly used by future developers?

Reply:
I think promoting `TM` as a friend of `LogicalCachedExtent` makes it challenging to see how many private members are used/dependent by `TM`, because all are exposed, although currently only `seen_by_users` is necessary.

Using the access methods, even if it may be wrongly used by accident, I think it is easier to check anytime by grepping the method name.

Or probably there is a better way I'm missing.